src/GoTrueClient.ts

```diff
@@ -530,28 +530,57 @@ export default class GoTrueClient {
   }
 
   /**
-   * Sets the session data from refresh token and returns current session or an error if the refresh token is invalid.
-   * @param refresh_token A refresh token returned by supabase auth.
+   * Sets the session data from the current session. If the current session is expired, setSession will take care of refreshing it to obtain a new session.
+   * If the refresh token in the current session is invalid and the current session has expired, an error will be thrown.
+   * If the current session does not contain at expires_at field, setSession will use the exp claim defined in the access token.
+   * @param currentSession The current session that minimally contains an access token, refresh token and a user.
    */
-  async setSession(refresh_token: string): Promise<AuthResponse> {
+  async setSession(refresh_token: string, access_token = ''): Promise<AuthResponse> {
     try {
-      if (!refresh_token) {
-        throw new AuthSessionMissingError()
-      }
-      const { data, error } = await this._refreshAccessToken(refresh_token)
-      if (error) {
-        return { data: { session: null, user: null }, error: error }
+      const timeNow = Date.now() / 1000
+      let expiresAt = timeNow
+      let hasExpired = true
+      let session: Session | null = null
+      if (access_token && access_token.split('.')[1]) {
+        const payload = JSON.parse(Buffer.from(access_token.split('.')[1], 'base64').toString())
```

Comment:
Buffer does not exist on the browser. The only way to parse Base64 on the browser is [btoa](https://developer.mozilla.org/en-US/docs/Web/API/btoa). You can use the code from the v1 PR.

Reply:
Also note that `btoa` does not accept Base64-URL, which is what JWT uses, so you need to replace the characters. You don't need to add padding (`=`) though.